00_Base/src/mapper/BaseTransactionMapper.ts

```diff
@@ -0,0 +1,141 @@
+    protected tariffRepository: SequelizeTariffRepository,
+    protected tariffsService: TariffsService,
+  ) {
+  }
+
+  protected async getLocationDTOsForTransactions(
+    transactions: Transaction[],
+  ): Promise<Map<string, LocationDTO>> {
+    const transactionIdToLocationMap: Map<string, LocationDTO> = new Map();
+    for (const transaction of transactions) {
+      const chargingStation = await transaction.$get('station');
+      if (!chargingStation || !chargingStation.locationId) {
+        continue;
+      }
+
+      const location = await this.locationsService.getLocationById(chargingStation.locationId);
+
+      if (!location || !location.data) {
+        continue;
+      }
+
+      transactionIdToLocationMap.set(transaction.transactionId, location.data);
+    }
+    return transactionIdToLocationMap;
+  }
+
+  protected async getTokensForTransactions(
+    transactions: Transaction[]
+  ): Promise<{ [key: string]: TokenDTO }> {
+    const transactionIdToTokenMap: { [transactionId: string]: TokenDTO } = {};
+
+    for (const transaction of transactions) {
+      const startEvent = transaction.transactionEvents?.find(
+        (event) => event.eventType === 'Started'
+      );
+      if (startEvent?.idToken) {
+        const tokenDto = await this.tokensRepository.getTokenDtoByIdToken(
+          startEvent.idToken.idToken,
+          startEvent.idToken.type
+        );
+        if (tokenDto) {
+          transactionIdToTokenMap[transaction.transactionId] = tokenDto;
+        }
+      }
+    }
+
+    return transactionIdToTokenMap;
+  }
+
+  protected async getTariffsForTransactions(
+    transactions: Transaction[]
+  ): Promise<Map<string, Tariff>> {
+    const transactionIdToTariffMap = new Map<string, Tariff>();
+    const uniqueStationIds = [...new Set(transactions.map(t => t.stationId))];
+
+    const stationIdToTariffMap = new Map<string, Tariff>();
+    await Promise.all(
+      uniqueStationIds.map(async (stationId) => {
+        const tariff = await this.tariffRepository.findByStationId(stationId);
+        if (tariff) {
+          stationIdToTariffMap.set(stationId, tariff);
+        }
+      })
+    );
+
+    for (const transaction of transactions) {
+      const tariff = stationIdToTariffMap.get(transaction.stationId);
+      if (tariff) {
+        transactionIdToTariffMap.set(transaction.transactionId, tariff);
+      }
+    }
+
+    return transactionIdToTariffMap;
+  }
+
+  protected async getOcpiTariffsForTransactions(
+    sessions: Session[],
+    transactionIdToTariffMap: Map<string, Tariff>,
+  ): Promise<Map<string, OcpiTariff>> {
+    const transactionIdToOcpiTariffMap = new Map<string, OcpiTariff>();
+    await Promise.all(
+      sessions.filter(session => transactionIdToTariffMap.get(session.id)).map(async (session) => {
+        const tariffKey = {
+          id: String(transactionIdToTariffMap.get(session.id)?.id),
+          // TODO: Ensure CPO Country Code, Party ID exists for the tariff in question
```

Comment:
yes--in a multi-tenant system, not all tenants might require OCPI data

Reply:
Is there a solution here in that case?

Presumably, if we are referencing a Tariff that does not have country code + party ID, then it is a non-OCPI Tariff, and there's not very much we can do except return the Session without Tariff information which seems problematic because we can then not compute cost.

Reply:
This is a good point. I think the validation should happen elsewhere... it should be impossible for a connector to lack a tariffId, and it should be impossible for a tariffId on a connector to not be an ocpi tariff (or for an ocpi tariff to be 'downgraded' to a non-ocpi tariff). If these things are true, this logic should never break.